rl4co/models/zoo/deepaco/antsystem.py

```diff
@@ -0,0 +1,245 @@
+        """Run the Ant System algorithm for a specified number of iterations.
+
+        Args:
+            td_initial: Initial state of the problem.
+            env: Environment representing the problem.
+            n_iterations: Number of iterations to run the algorithm.
+
+        Returns:
+            td: The final state of the problem.
+            actions: The final actions chosen by the algorithm.
+            reward: The final reward achieved by the algorithm.
+        """
+        for _ in range(n_iterations):
+            # reset environment
+            td: TensorDict = env.reset(td_initial.clone(recurse=False))  # type: ignore
+            self.one_step(td, env)
+
+        td, env = self._recreate_final_routes(td_initial, env)
+
+        assert self.final_actions is not None and self.final_reward is not None
+        return td, self.final_actions, self.final_reward
+
+    def one_step(self, td: TensorDict, env: RL4COEnvBase):
+        """Run one step of the Ant System algorithm.
+
+        Args:
+            td: Current state of the problem.
+            env: Environment representing the problem.
+
+        Returns:
+            actions: The actions chosen by the algorithm.
+            reward: The reward achieved by the algorithm.
+        """
+        # sampling
+        actions, reward = self._sampling(td, env)
+        # local search, reserved for extensions
+        actions, reward = self._local_search(actions, reward)
+        # update final actions and rewards
+        self._update_results(actions, reward)
+        # update pheromone matrix
+        self._update_pheromone(actions, reward)
+
+        return actions, reward
+
+    def _sampling(
+        self,
+        td: TensorDict,
+        env: RL4COEnvBase,
+    ):
+        # Sample from heatmaps
+        # p = phe**alpha * heu**beta <==> log(p) = alpha*log(phe) + beta*log(heu)
+        heatmaps_logp = (
+            self.alpha * torch.log(self.pheromone) + self.beta * self.log_heuristic
+        )
+        self.decode_strategy = Sampling(multistart=True, num_starts=self.n_ants)
+        td, env, num_starts = self.decode_strategy.pre_decoder_hook(td, env)
+        while not td["done"].all():
+            log_p, mask = NARDecoder._get_log_p(td, heatmaps_logp, num_starts)
+            td = self.decode_strategy.step(log_p, mask, td)
+            td = env.step(td)["next"]
+
+        outputs, actions, td, env = self.decode_strategy.post_decoder_hook(td, env)
+        reward = env.get_reward(td, actions)
+
+        if self.require_logp:
+            self.all_records.append((outputs, actions, reward, td.get("mask", None)))
+
+        # reshape from (batch_size * n_ants, ...) to (batch_size, n_ants, ...)
+        reward = reward.view(self.batch_size, self.n_ants)
+        actions = actions.view(self.batch_size, self.n_ants, -1)
+
+        return actions, reward
+
+    def _local_search(self, actions: Tensor, reward: Tensor) -> Tuple[Tensor, Tensor]:
```

Comment:
Note: the current plan is to extend `_local_search` into an (optional) plug-in class method from the `Env` so we can fully modularize this part, which is the work  @hyeok9855 is doing as you can see in [this PR](https://github.com/ai4co/rl4co/pull/140) 

The signature would be something like:

```python
class CustomRL4COEnv()
   [...]
    def local_search(self, td: TensorDict, actions: torch.Tensor, **kwargs) -> torch.Tensor:
    """Input: td (state), actions
         Returns: improved actions
     """
```

I think this could be further modularized. For example, we could set a parameter for setting the local search operator (by default `None`). What do you think?

Reply:
I think that would be a great idea!